// astro/dist/content/vite-plugin-content-assets.js
import { pathToFileURL } from "url";
import { moduleIsTopLevelPage, walkParentInfos } from "../core/build/graph.js";
import { getPageDataByViteID } from "../core/build/internal.js";
import { createViteLoader } from "../core/module-loader/vite.js";
import { getStylesForURL } from "../core/render/dev/css.js";
import { getScriptsForURL } from "../core/render/dev/scripts.js";
import {
  contentFileExts,
  LINKS_PLACEHOLDER,
  PROPAGATED_ASSET_FLAG,
  SCRIPTS_PLACEHOLDER,
  STYLES_PLACEHOLDER
} from "./consts.js";
function isPropagatedAsset(viteId) {
  const url = new URL(viteId, "file://");
  return url.searchParams.has(PROPAGATED_ASSET_FLAG) && contentFileExts.some((ext) => url.pathname.endsWith(ext));
}
function astroContentAssetPropagationPlugin({ mode }) {
  let devModuleLoader;
  return {
    name: "astro:content-asset-propagation",
    enforce: "pre",
    configureServer(server) {
      if (mode === "dev") {
        devModuleLoader = createViteLoader(server);
      }
    },
    load(id) {
      if (isPropagatedAsset(id)) {
        const basePath = id.split("?")[0];
        const code = `
					export { Content, getHeadings, frontmatter } from ${JSON.stringify(basePath)};
					export const collectedLinks = ${JSON.stringify(LINKS_PLACEHOLDER)};
					export const collectedStyles = ${JSON.stringify(STYLES_PLACEHOLDER)};
					export const collectedScripts = ${JSON.stringify(SCRIPTS_PLACEHOLDER)};
				`;
        return { code };
      }
    },
    async transform(code, id, options) {
      var _a;
      if (!(options == null ? void 0 : options.ssr))
        return;
      if (devModuleLoader && isPropagatedAsset(id)) {
        const basePath = id.split("?")[0];
        if (!((_a = devModuleLoader.getModuleById(basePath)) == null ? void 0 : _a.ssrModule)) {
          await devModuleLoader.import(basePath);
        }
        const { stylesMap, urls } = await getStylesForURL(
          pathToFileURL(basePath),
          devModuleLoader,
          "development"
        );
        const hoistedScripts = await getScriptsForURL(pathToFileURL(basePath), devModuleLoader);
        return {
          code: code.replace(JSON.stringify(LINKS_PLACEHOLDER), JSON.stringify([...urls])).replace(JSON.stringify(STYLES_PLACEHOLDER), JSON.stringify([...stylesMap.values()])).replace(JSON.stringify(SCRIPTS_PLACEHOLDER), JSON.stringify([...hoistedScripts]))
        };
      }
    }
  };
}
function astroContentProdBundlePlugin({ internals }) {
  return {
    name: "astro:content-prod-bundle",
    async generateBundle(_options, bundle) {
      var _a;
      for (const [_, chunk] of Object.entries(bundle)) {
        if (chunk.type === "chunk" && chunk.code.includes(LINKS_PLACEHOLDER)) {
          for (const id of Object.keys(chunk.modules)) {
            for (const [pageInfo, depth, order] of walkParentInfos(id, this)) {
              if (moduleIsTopLevelPage(pageInfo)) {
                const pageViteID = pageInfo.id;
                const pageData = getPageDataByViteID(internals, pageViteID);
                if (!pageData)
                  continue;
                const entryCss = (_a = pageData.contentCollectionCss) == null ? void 0 : _a.get(id);
                if (!entryCss)
                  continue;
                chunk.code = chunk.code.replace(
                  JSON.stringify(LINKS_PLACEHOLDER),
                  JSON.stringify([...entryCss])
                );
              }
            }
          }
        }
      }
    }
  };
}
export {
  astroContentAssetPropagationPlugin,
  astroContentProdBundlePlugin
};
